text/3738-drop-type-destructuring.md

```diff
@@ -0,0 +1,368 @@
+- Feature Name: `drop_type_destructuring`
+- Start Date: 2024-12-08
+- RFC PR: [rust-lang/rfcs#3738](https://github.com/rust-lang/rfcs/pull/3738)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Rust does not allow destructuring types which implement the `Drop` trait. This means that moving data out of such types is hard and error prone. The rationale is that once fields are moved out, the type's `Drop` implementation cannot run, which can be undesired. This RFC proposes to allow destructuring anyway, in certain situations.
```

Comment:
Could you add more detail here? The reasons for this language quirk are crucial to this RFC, what exactly is it trying to prevent?

Reply:
Consider this example:
```rust
let x = X { y: Y }; // X: Drop, Y: !Copy
let y = x.y;
// implicitly this line contains `drop(x)`
```
The [implicit `drop(x)` would be an error](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=5c7a7c04e1c9fab72d743dc81321d473), because `x.y` is moved out. So the "you can't move fields out of `Drop` types" is a specialization of "use of partially moved value".

We _could_ say that `let y = x.y` disables the drop, but that is a footgun, see https://github.com/rust-lang/rfcs/pull/3738#discussion_r1874953956.

Do you think we should write this more clearly in the RFC?

Reply:
> Do you think we should write this more clearly in the RFC?

Very much so. My mental model makes me expect `let y = x.y` to disable the drop, and it's not obvious to me that this is a footgun, so I'd appreciate a motivation + example for why we don't allow this.

Reply:
> So the "you can't move fields out of Drop types" is a specialization of "use of partially moved value".

Except it is more strict, because moving fields out of `Drop` types is forbidden even if you put something back afterwards:

```rs
// only allowed if `Foo` does NOT implement Drop
let mut foo = Foo {
    string: "old value".to_string(),
};
drop(foo.string);
foo.string = "new value".to_string();
drop(foo);
```

But this could be allowed for `Drop` types as well.

Reply:
> Consider this example:
> 
> ```rust
> let x = X { y: Y }; // X: Drop, Y: !Copy
> let y = x.y;
> // implicitly this line contains `drop(x)`
> ```
> 
> The [implicit `drop(x)` would be an error](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=5c7a7c04e1c9fab72d743dc81321d473), because `x.y` is moved out. So the "you can't move fields out of `Drop` types" is a specialization of "use of partially moved value".
> 
> We _could_ say that `let y = x.y` disables the drop, but that is a footgun, see [#3738 (comment)](https://github.com/rust-lang/rfcs/pull/3738#discussion_r1874953956).
> 
> Do you think we should write this more clearly in the RFC?

Is `let X { y } = x` is disallowed when `X: Drop`, `Y: !Copy` simply because it would be weird and seemingly inconsistent for `let X { y } = x` to be legal while `let y = x.y` is illegal? Or is there a deeper reason?

The reason I ask is that, as discussed in [this thread](https://github.com/rust-lang/rfcs/pull/3738#discussion_r1874719330), an alternative design would just be to permit `let X { y } = x`, and have it semantically move all of the fields rather than dropping the original `x`. That option would still permit `let y = x.y` to remain illegal.

Reply:
The left side of a let binding is an irrefutable pattern, so proposing this is proposing that some specific kinds of irrefutable pattern are now special.  That's the deeper reason to me.

Also, making this specific syntax work can cause code to change semantics rather than producing compile errors.  Today, if the type being destructured suddenly gains a Drop impl, then your code stops compiling.  If some crate `dep` adds a Drop impl tomorrow, then your code needed to decide what to do--but you used the magic syntax by accident.

Reply:
Makes sense, thanks for the explanation!

Reply:
> Also, making this specific syntax work can cause code to change semantics rather than producing compile errors. Today, if the type being destructured suddenly gains a Drop impl, then your code stops compiling. If some crate dep adds a Drop impl tomorrow, then your code needed to decide what to do--but you used the magic syntax by accident.

"change semantics" is a bit strong: it's still just destructuring a type. If you don't want the user to destructure your type, make the fields private. Either way adding a `Drop` impl will stay a breaking change because it affects borrow-checking in other ways.

Reply:
I still don't understand in what cases ths obvious "just let the user destructure any type" would be a footgun. Could someone provide concrete examples of types that have both a significant `Drop` impl and public fields where this could cause issues? I've never personally encountered such a case.

Reply:
from a little searching, I found a [unit type with a `Drop` impl](https://github.com/sarvex/sprocketnes/blob/e8f8b798155b3e586f4ec5d0ef26c64e31be7cc0/audio.rs#L124), if just destructuring using normal syntax was allowed, then writing `let AudioLock = the_lock;` or `let AudioLock {} = the_lock;` would be a foot-gun because it wouldn't disable `Drop` because all fields are `Copy` (vacuously true).

Reply:
actually, pattern matching a unit type with a `Drop` impl is [already a footgun](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d3eeb2463e9f629a65e4810a63546509)

Reply:
Thanks for the example! I think there are two footguns at play:
1. users writing `let <pat> = <drop_val>;` and expecting that to disable `Drop`;
2. users holding a value with an important `Drop` impl and accidentally disabling `Drop` by pattern-matching on it.

My understanding of the RFC is that is argues against "just let destructuring work" because that would create footgun 2. I am still skeptical that this is a real issue and open to seeing examples of that. What you just showed is footgun 1, which already exists and arguably would be made worse if we "just let destructuring work"?

Reply:
Yes, this is a real issue. I can't show you the code that causes the problem because it doesn't exist yet, as we don't permit it in the actual common cases. But there are cases where something has a meaningful Drop impl but a certain function is *intentionally* supposed to disable that Drop, and so it is problematic if the specially marked operation is not, er, specially marked.

Allocated container types, for instance.